Mathlib/Tactic/Widget/Calc.lean

```diff
@@ -124,19 +126,30 @@ def CalcPanel : Component CalcParams :=
   mk_rpc_widget% CalcPanel.rpc
 
 namespace Lean.Elab.Tactic
-open Meta
+
+open Lean Meta Tactic TryThis in
+/-- Create a `calc` proof. -/
+elab stx:"calc?" : tactic => do
+  let some calcRange := (← getFileMap).rangeOfStx? stx | throwError "calc? failed"
+  let indent := calcRange.start.character + 2
+  let spc := String.replicate indent ' '
+  let goalType ← getMainTarget
+  unless (← Lean.Elab.Term.getCalcRelation? goalType).isSome do
+    throwError "Cannot start a calculation here: the goal{indentExpr goalType}\nis not a relation."
+  let goalFmt ← Meta.ppExpr (← getMainTarget)
```

Comment:
This presumably needs a `withMainContext` around it (and perhaps just the whole tactic) to avoid printing issues.

Reply:
Weirdly, it worked with this, but I added it just in case.

Reply:
I would expect it to break if you inserted a `have` into the context or something similar. If the tactic is first in `by`, then `withMainContext` is a no-op.